src/MSBuild/ValidateMSBuildPackageDependencyVersions.cs

```diff
@@ -0,0 +1,87 @@
+                                            string version = string.Empty;
+                                            foreach (var dependentAssembly in assemblyBindingXmlElement.ChildNodes)
+                                            {
+                                                if (dependentAssembly is XmlElement dependentAssemblyXmlElement)
+                                                {
+                                                    if (dependentAssemblyXmlElement.Name.Equals("assemblyIdentity", StringComparison.OrdinalIgnoreCase))
+                                                    {
+                                                        foreach (var assemblyIdentityAttribute in dependentAssemblyXmlElement.Attributes)
+                                                        {
+                                                            if (assemblyIdentityAttribute is XmlAttribute assemblyIdentityAttributeXmlElement && assemblyIdentityAttributeXmlElement.Name.Equals("name", StringComparison.OrdinalIgnoreCase))
+                                                            {
+                                                                name = assemblyIdentityAttributeXmlElement.Value;
+                                                            }
+                                                        }
+                                                    }
+                                                    else if (dependentAssemblyXmlElement.Name.Equals("bindingRedirect", StringComparison.OrdinalIgnoreCase))
+                                                    {
+                                                        foreach (var bindingRedirectAttribute in dependentAssemblyXmlElement.Attributes)
+                                                        {
+                                                            if (bindingRedirectAttribute is XmlAttribute bindingRedirectAttributeXmlElement && bindingRedirectAttributeXmlElement.Name.Equals("newVersion", StringComparison.OrdinalIgnoreCase))
+                                                            {
+                                                                version = bindingRedirectAttributeXmlElement.Value;
+                                                            }
+                                                        }
+                                                    }
+                                                }
+                                            }
+                                            if (!string.IsNullOrEmpty(name) && !string.IsNullOrEmpty(version))
+                                            {
+                                                string path = Path.Combine(AssemblyPath, name + ".dll");
+                                                string assemblyVersion = File.Exists(path) ? Assembly.LoadFile(path).GetName().Version.ToString() : version;
```

Comment:
Calling `LoadFile` will load the assemblies into the running MSBuild process in a way that's not unloadable, potentially causing problems on subsequent builds (the same problem you're tackling for task assemblies in #7132).

Fortunately since we just need the assembly identity, we don't need to use `System.Reflection.Metadata` here.

```suggestion
                                                string assemblyVersion = AssemblyName.GetAssemblyName(path).Version.ToString();
```

I don't think the `File.Exists` default is right--we want to guarantee that we ship the file if we have a binding redirect for it.

Reply:
I've found a number of cases that seem pretty reasonable that we should ignore like M.B.Conversion.Core, M.B.Engine, and M.NET.StringTools.net35. I can make a list of assemblies to ignore, but it's a bit messy.

Reply:
I'll ignore all of them, then let you comment on which to keep. (So far, I'd say those three and none of the others.)